src/internal.c

```diff
@@ -7530,9 +7534,12 @@ void FreeKeyExchange(WOLFSSL* ssl)
 
     /* Cleanup digest buffer */
     if (ssl->buffers.digest.buffer) {
-        XFREE(ssl->buffers.digest.buffer, ssl->heap, DYNAMIC_TYPE_DIGEST);
+        if(!ssl->options.dontFreeDigest) {
+            XFREE(ssl->buffers.digest.buffer, ssl->heap, DYNAMIC_TYPE_DIGEST);
+        }
         ssl->buffers.digest.buffer = NULL;
```

Comment:
If the buffer isn't freed, then you shouldn't set the buffer to NULL?
Also for other changes below.

Reply:
When SetDigest is called, the digest buffer points to a non-malloc’ed area. In this case, it is correct to simply unset the pointer and the flag.  Hopefully this is the case for the others below as well.  Let me know if this resolves your question.

Reply:
Yes, it's ok. It just looks always suspicious setting a pointer to NULL without the free before.
Might be more obvious if the name "dontFreeDigest" would not describe what, but why "I do not own the buffer".

Reply:
Maybe a clarifying comment could help the situation?

Reply:
Added comment to help clarify.